File: OrderMatchingEngine/src/main/java/com/ndgndg91/ordermatchingengine/order/OrderBook.java
package com.ndgndg91.ordermatchingengine.order;

import com.ndgndg91.ordermatchingengine.global.OrderServiceException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;

import java.util.*;
import java.util.concurrent.ConcurrentSkipListSet;

@Slf4j
public class OrderBook {
    private final Symbol symbol;
    private final SortedSet<OrderEntry> limitBids;
    private final SortedSet<OrderEntry> limitAsks;
    private final SortedSet<OrderEntry> marketBids;
    private final SortedSet<OrderEntry> marketAsks;

    public OrderBook(Symbol symbol) {
        this.symbol = symbol;
        this.limitBids = new ConcurrentSkipListSet<>((e1, e2) -> {
            int c = e2.getPrice().compareTo(e1.getPrice());
            if (c == 0) return e1.getTimestamp().compareTo(e2.getTimestamp());
            return c;
        });
        this.limitAsks = new ConcurrentSkipListSet<>((e1, e2) -> {
            int c = e1.getPrice().compareTo(e2.getPrice());
            if (c == 0) return e1.getTimestamp().compareTo(e2.getTimestamp());
            return c;
        });
        this.marketBids = new ConcurrentSkipListSet<>(Comparator.comparing(OrderEntry::getTimestamp));
        this.marketAsks = new ConcurrentSkipListSet<>(Comparator.comparing(OrderEntry::getTimestamp));
    }

    private SortedSet<OrderEntry> selectQueue(PriceType priceType, OrderType orderType) {
        switch (priceType) {
            case MARKET:
                return marketQueue(orderType);
            case LIMIT:
                return limitQueue(orderType);
            default:
                throw new OrderServiceException(null, HttpStatus.BAD_REQUEST.value(), "not found appropriate queue");
        }
    }

    private SortedSet<OrderEntry> limitQueue(OrderType orderType) {
        switch (orderType) {
            case BID:
                return this.limitBids;
            case ASK:
                return this.limitAsks;
            default:
                throw new OrderServiceException(null, HttpStatus.BAD_REQUEST.value(), "not found appropriate queue");
        }
    }

    private SortedSet<OrderEntry> marketQueue(OrderType orderType) {
        switch (orderType) {
            case BID:
                return this.marketBids;
            case ASK:
                return this.marketAsks;
            default:
                throw new OrderServiceException(null, HttpStatus.BAD_REQUEST.value(), "not found appropriate queue");
        }
    }

    public void addOrder(Order order) {
        SortedSet<OrderEntry> queue = selectQueue(order.getPriceType(), order.getOrderType());
        if (queue.stream().parallel().anyMatch(e -> e.getOrderId().equals(order.getOrderId()))) {
            log.info("already exists, failed to add {}", order);
        } else {
            OrderEntry e = new OrderEntry(order);
            queue.add(e);
            log.info("success to add {}", order);
        }
    }

    public void modifyOrder(Order order) {
        SortedSet<OrderEntry> queue = selectQueue(order.getPriceType(), order.getOrderType());
        OrderEntry oe = queue.stream()
                .parallel()
                .filter(e -> e.getOrderId().equals(order.getOrderId()))
                .findFirst()
                .orElseThrow(() -> new OrderServiceException(null, HttpStatus.BAD_REQUEST.value(), "when modifying order, not found order " + order.getOrderId()));
        queue.remove(oe);
        queue.add(new OrderEntry(order));
    }

    public void cancelOrder(Order order) {
        SortedSet<OrderEntry> queue = selectQueue(order.getPriceType(), order.getOrderType());

        OrderEntry target = queue.stream()
                .parallel()
                .filter(e -> e.getOrderId().equals(order.getOrderId()))
                .findFirst()
                .orElseThrow(() -> new OrderServiceException(null, HttpStatus.BAD_REQUEST.value(), "when canceling order, not found order : " + order.getOrderId()));
        queue.remove(target);
    }

    public Optional<OrderEntry> bidsPoll() {
        if (!this.marketBids.isEmpty()) {
            return Optional.ofNullable(poll(this.marketBids));
        }

        return Optional.ofNullable(poll(this.limitBids));
    }

    public Optional<OrderEntry> asksPoll() {
        if (!this.marketAsks.isEmpty()) {
            return Optional.ofNullable(poll(this.marketAsks));
        }

        return Optional.ofNullable(poll(this.limitAsks));
    }

    public MatchResult match(PriceType priceType, OrderType orderType) {
        switch (priceType) {
            case LIMIT:
                return matchLimitOrder();
            case MARKET:
                return matchMarketOrder(orderType);
            default:
                // must do not happen
                return null;
        }
    }

    private MatchResult matchLimitOrder() {
        OrderEntry bid = peek(this.limitBids);
        OrderEntry ask = peek(this.limitAsks);
        if (ask == null || bid == null) {
            return null;
        }

        int c = bid.getPrice().compareTo(ask.getPrice());
        int d = bid.shares() - ask.shares();
        // matched
        if (c >= 0) {
            if (d == 0) { // totally matched
                poll(this.limitBids);
                poll(this.limitAsks);
                return MatchResult.exact(bid, symbol, ask);
            } else if (d > 0) { // need more ask
                List<OrderEntry> tAsks = new ArrayList<>();
                tAsks.add(poll(this.limitAsks));
                while (d > 0) {
                    OrderEntry peek = peek(this.limitAsks);
                    if (peek == null || bid.getPrice().compareTo(peek.getPrice()) < 0) {
                        limitAsks.addAll(tAsks);
                        return null;
                    }

                    if (d < peek.shares()) {
                        peek.partialMatched(bid, d);
                        tAsks.add(peek);
                        poll(this.limitBids);
                        d = 0;
                    } else if (d > peek.shares()) {
                        d -= peek.shares();
                        tAsks.add(poll(this.limitAsks));
                    } else { // d == peek.shares()
                        tAsks.add(poll(this.limitAsks));
                        break;
                    }
                }

                return MatchResult.bigBid(bid, symbol, tAsks);
            } else { // ask has more shares
                ask.partialMatched(bid);
                poll(this.limitBids);
                return MatchResult.bigAsk(bid, symbol, ask);
            }
        } else { // not matched
            return null;
        }
    }

    private MatchResult matchMarketOrder(OrderType orderType) {
        switch (orderType) {
            case BID:
                return matchMarketBidOrder();
            case ASK:
                return matchMarketAskOrder();
            default:
                // must do not happen
                return null;
        }
    }

    private MatchResult matchMarketBidOrder() {
        OrderEntry mBid = peek(this.marketBids);
        OrderEntry lAsk = peek(this.limitAsks);
        if (mBid == null || lAsk == null) {
            return null;
        }

        int d = mBid.shares() - lAsk.shares();
        if (d == 0) { // exact
            poll(this.marketBids);
            poll(this.limitAsks);
            return MatchResult.exact(mBid, symbol, lAsk);
        } else if (d > 0) { // need more ask
            List<OrderEntry> tAsks = new ArrayList<>();
            tAsks.add(poll(this.limitAsks));
            while (d > 0) {
                OrderEntry peek = peek(this.limitAsks);
                if (peek == null) {
                    limitAsks.addAll(tAsks);
                    return null;
                }

                if (d < peek.shares()) {
                    peek.partialMatched(mBid, d);
                    tAsks.add(peek);
                    poll(this.marketBids);
                    d = 0;
                } else if (d > peek.shares()) {
                    d -= peek.shares();
                    tAsks.add(poll(this.limitAsks));
                } else { // d == peek.shares()
                    tAsks.add(poll(this.limitAsks));
                    break;
                }
            }

            return MatchResult.bigBid(mBid, symbol, tAsks);
        } else { // ask partial match
            lAsk.partialMatched(mBid);
            poll(this.marketBids);
            return MatchResult.bigAsk(mBid, symbol, lAsk);
        }
    }

    private MatchResult matchMarketAskOrder() {
        OrderEntry mAsk = peek(this.marketAsks);
        OrderEntry lBid = peek(this.limitBids);
        if (mAsk == null || lBid == null) {
            return null;
        }

        int d = mAsk.shares() - lBid.shares();
        if (d == 0) {
            poll(this.marketAsks);
            poll(this.limitBids);
            return MatchResult.exact(mAsk, symbol, lBid);
        } else if (d > 0) { // need more bid
            List<OrderEntry> tBids = new ArrayList<>();
            tBids.add(poll(this.limitBids));
            while (d > 0) {
                OrderEntry peek = peek(this.limitBids);
                if (peek == null) {
                    limitBids.addAll(tBids);
                    return null;
                }

                if (d < peek.shares()) {
                    peek.partialMatched(mAsk, d);
                    tBids.add(peek);
                    poll(this.marketBids);
                    d = 0;
                } else if (d > peek.shares()) {
                    d -= peek.shares();
                    tBids.add(poll(this.limitBids));
                } else { // d == peek.shares()
                    tBids.add(poll(this.limitBids));
                    break;
                }
            }

            return MatchResult.bigBid(mAsk, symbol, tBids);
        } else { // bid partial match
            lBid.partialMatched(mAsk);
            poll(this.marketAsks);
            return MatchResult.bigAsk(mAsk, symbol, lBid);
        }
    }

    private OrderEntry peek(SortedSet<OrderEntry> queue) {
        try {
            return queue.first();
        } catch (NoSuchElementException e) {
            return null;
        }
    }

    private OrderEntry poll(SortedSet<OrderEntry> queue) {
        try {
            OrderEntry first = queue.first();
            queue.remove(first);
            return first;
        } catch (NoSuchElementException e) {
            return null;
        }
    }
}

File: OrderMatchingEngineKotlin/src/main/kotlin/com/ndgndg91/ordermatchingenginekotlin/order/OrderBook.kt
package com.ndgndg91.ordermatchingenginekotlin.order

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.ConcurrentSkipListSet
import kotlin.NoSuchElementException
import kotlin.collections.ArrayList

fun SortedSet<OrderEntry>.firstByOrderId(id: String): OrderEntry? = try {
    this.first { it.orderId == id }
} catch (e: NoSuchElementException) {
    null
}


class OrderBook(private val symbol: Symbol) {
    private val log: Logger = LoggerFactory.getLogger(OrderBook::class.java)

    private val limitBids: SortedSet<OrderEntry> = ConcurrentSkipListSet { e1, e2 ->
        val c = e2.price.compareTo(e1.price)
        if (c == 0) e1.timestamp.compareTo(e2.timestamp) else c
    }
    private val limitAsks: SortedSet<OrderEntry> = ConcurrentSkipListSet { e1, e2 ->
        val c = e1.price.compareTo(e2.price)
        if (c == 0) e1.timestamp.compareTo(e2.timestamp) else c
    }

    private val marketBids: SortedSet<OrderEntry> = ConcurrentSkipListSet { e1, e2 ->
        e1.timestamp.compareTo(e2.timestamp)
    }
    private val marketAsks: SortedSet<OrderEntry> = ConcurrentSkipListSet { e1, e2 ->
        e1.timestamp.compareTo(e2.timestamp)
    }

    private fun selectQueue(priceType: PriceType, orderType: OrderType): SortedSet<OrderEntry> {
        return when (priceType) {
            PriceType.MARKET -> marketQueue(orderType)
            PriceType.LIMIT -> limitQueue(orderType)
        }
    }

    private fun marketQueue(orderType: OrderType): SortedSet<OrderEntry> {
        return when (orderType) {
            OrderType.BID -> this.marketBids
            OrderType.ASK -> this.marketAsks
        }
    }

    private fun limitQueue(orderType: OrderType): SortedSet<OrderEntry> {
        return when (orderType) {
            OrderType.BID -> this.limitBids
            OrderType.ASK -> this.limitAsks
        }
    }

    fun addOrder(order: Order) {
        val queue = selectQueue(order.priceType, order.orderType)
        if (kotlin.runCatching { queue.first { it.orderId == order.orderId } }.isSuccess) {
            log.info("already exists, failed to add $order")
            return
        } else {
            val e = order.toOrderEntry()
            queue.add(e)
            log.info("success to add $order")
        }
    }

    fun modifyOrder(order: Order) {
        val queue = selectQueue(order.priceType, order.orderType)
        kotlin.runCatching { queue.first { it.orderId == order.orderId } }
            .onSuccess { queue.remove(it) }.also { log.info("success to remove origin : ${it.getOrThrow()}") }
        val e = order.toOrderEntry()
        queue.add(e).also { log.info("success to modify : $e") }
    }

    fun cancelOrder(order: Order) {
        val queue = selectQueue(order.priceType, order.orderType)

        kotlin.runCatching { queue.first { it.orderId == order.orderId } }
            .onSuccess { queue.remove(it) }
    }

    fun find(orderType: OrderType, orderId: String): OrderEntry? = when (orderType) {
        OrderType.ASK -> {
            this.limitAsks.firstByOrderId(orderId)?: this.marketAsks.firstByOrderId(orderId)
        }
        OrderType.BID -> {
            this.limitBids.firstByOrderId(orderId)?: this.marketBids.firstByOrderId(orderId)
        }
    }


    fun bidsPoll(): OrderEntry? = try {
        if (!this.marketBids.isEmpty()) {
            poll(this.marketBids)
        } else {
            poll(this.limitBids)
        }
    } catch (e: NoSuchElementException) {
        null
    }


    fun asksPoll(): OrderEntry? = try {
        if (!this.marketAsks.isEmpty()) {
            poll(this.marketAsks)
        } else {
            poll(this.limitAsks)
        }
    } catch (e: NoSuchElementException) {
        null
    }

    fun match(priceType: PriceType, orderType: OrderType): List<MatchResult> {
        return when (priceType) {
            PriceType.LIMIT -> matchLimitOrder(orderType)
            PriceType.MARKET -> matchMarketOrder(orderType)
        }
    }

    private fun matchLimitOrder(orderType: OrderType): List<MatchResult> = when(orderType) {
        OrderType.BID -> matchLimitOrder()
        OrderType.ASK -> matchLimitOrder()
    }

    private fun matchLimitOrder(): List<MatchResult> {
        val bid = peek(this.limitBids)
        val ask = peek(this.limitAsks)
        if (ask == null || bid == null) {
            return listOf()
        }

        val c = bid.price.compareTo(ask.price)
        var d = bid.shares() - ask.shares()
        // matched
        if (c >= 0) {
            if (d == 0) { // exact matched
                poll(this.limitBids)
                poll(this.limitAsks)
                return listOf(MatchResult.bidExactAsk(bid, symbol, ask))
            } else if (d > 0) { // bid has more shares and need more ask : partial matched
                val tAsks = ArrayList<OrderEntry>()
                tAsks.add(poll(this.limitAsks)!!)
                while (d > 0) {
                    val peek = peek(this.limitAsks)
                    if (peek == null || bid.price < peek.price) {
                        return tAsks
                            .map { bid.partialMatched(it);MatchResult.smallAskBigBid(it, symbol, bid) }
                            .toList()
//                        this.limitAsks.addAll(tAsks)
//                        return null
                    }

                    if (d < peek.shares()) {
                        peek.partialMatched(bid, d)
                        tAsks.add(peek)
                        poll(this.limitBids)
                        d = 0
                    } else if (d > peek.shares()) {
                        d -= peek.shares()
                        tAsks.add(poll(this.limitAsks)!!)
                    } else {
                        tAsks.add(poll(this.limitAsks)!!)
                        break
                    }
                }

                return listOf(MatchResult.bigBidSmallAsks(bid, symbol, tAsks))
            } else { // ask has more shares and need more bid : partial matched
                ask.partialMatched(bid)
                poll(this.limitBids)
                return listOf(MatchResult.smallBidBigAsk(bid, symbol, ask))
            }
        } else { // not matched
            return listOf()
        }
    }

    private fun matchMarketOrder(orderType: OrderType): List<MatchResult> {
        return when (orderType) {
            OrderType.BID -> matchMarketBidOrder()
            OrderType.ASK -> matchMarketAskOrder()
        }
    }

    private fun matchMarketBidOrder(): List<MatchResult> {
        val mBid = peek(this.marketBids)
        val lAsk = peek(this.limitAsks)
        if (mBid == null || lAsk == null) {
            return listOf()
        }

        var d = mBid.shares() - lAsk.shares()
        if (d == 0) { // exact matched
            poll(this.marketBids)
            poll(this.limitAsks)
            return listOf(MatchResult.bidExactAsk(mBid, symbol, lAsk))
        } else if (d > 0) { // market bid has more shares and need more limit ask
            val tAsks = ArrayList<OrderEntry>()
            tAsks.add(poll(this.limitAsks)!!)
            while (d > 0) {
                val peek = peek(this.limitAsks)
                if (peek == null) {
                    return tAsks
                        .map { mBid.partialMatched(it);MatchResult.smallAskBigBid(it, symbol, mBid) }
                        .toList()
//                    this.limitAsks.addAll(tAsks)
//                    return null
                }

                if (d < peek.shares()) {
                    peek.partialMatched(mBid, d)
                    tAsks.add(peek)
                    poll(this.marketBids)
                    d = 0
                } else if (d > peek.shares()) {
                    d -= peek.shares()
                    tAsks.add(poll(this.limitAsks)!!)
                } else {
                    tAsks.add(poll(this.limitAsks)!!)
                    poll(this.marketBids)
                    break
                }
            }

            return listOf(MatchResult.bigBidSmallAsks(mBid, symbol, tAsks))
        } else { // limit ask has more shares and need more market bid
            lAsk.partialMatched(mBid)
            poll(this.marketBids)
            return listOf(MatchResult.smallBidBigAsk(mBid, symbol, lAsk))
        }
    }

    private fun matchMarketAskOrder(): List<MatchResult> {
        val mAsk = peek(this.marketAsks)
        val lBid = peek(this.limitBids)
        if (mAsk == null || lBid == null) {
            return listOf()
        }

        var d = mAsk.shares() - lBid.shares()
        if (d == 0) { // exact matched
            poll(this.marketAsks)
            poll(this.limitBids)
            return listOf(MatchResult.bidExactAsk(mAsk, symbol, lBid))
        } else if (d > 0) { // ask has more shares and need more bid
            val tBids = ArrayList<OrderEntry>()
            tBids.add(poll(this.limitBids)!!)
            while (d > 0) {
                val peek = peek(this.limitBids)
                if (peek == null) {
                    return tBids
                        .map { mAsk.partialMatched(it);MatchResult.smallBidBigAsk(it, symbol, mAsk) }
                        .toList()
//                    this.limitBids.addAll(tBids)
//                    return null
                }

                if (d < peek.shares()) {
                    peek.partialMatched(mAsk, d)
                    tBids.add(peek)
                    poll(this.marketBids)
                    d = 0
                } else if (d > peek.shares()) {
                    d -= peek.shares()
                    tBids.add(poll(this.limitBids)!!)
                } else { // d == peek.shares()
                    tBids.add(poll(this.limitBids)!!)
                    poll(this.marketAsks)
                    break
                }
            }

            return listOf(MatchResult.bigBidSmallAsks(mAsk, symbol, tBids))
        } else { // bid has more shares and need more ask
            lBid.partialMatched(mAsk)
            poll(this.marketAsks)
            return listOf(MatchResult.smallBidBigAsk(mAsk, symbol, lBid))
        }
    }

    private fun peek(queue: SortedSet<OrderEntry>): OrderEntry? {
        return try {
            queue.first()
        } catch (e: NoSuchElementException) {
            null
        }
    }

    private fun poll(queue: SortedSet<OrderEntry>): OrderEntry? {
        return try {
            val first = queue.first()
            queue.remove(first)
            first
        } catch (e: NoSuchElementException) {
            null
        }
    }

}